### PR TITLE
ext token ttl mishandling of edge cases when comparing

### PR DESCRIPTION
**Pull request by andreas-kupries**
## Issue: 

No issue. During the discussion about https://github.com/rancher/rancher/issues/49588 it was realized that the
comparison of TTL values forgot to properly handle special cases (infinity, default).

## Problem

See above
 
## Solution

A custom comparison function is used to handle all edge cases properly.
 
## Testing

Added unit tests for the new comparator.

### Automated Testing

* Test types added/modified:
    * Unit
    * Integration (Go Framework)
    * Integration (v2prov Framework)
    * Validation (Go Framework)
    * Other - Explain: _EXPLAIN_
    * None
    * _REMOVE NOT APPLICABLE BULLET POINTS ABOVE_
* If "None" - Reason: _EXPLAIN THE REASON_
  
* If "None" - GH Issue/PR: _LINK TO GH ISSUE/PR TO ADD TESTS_

Summary: _TODO_

### Regressions Considerations

_TODO_

Existing / newly added automated tests that provide evidence there are no regressions:
* _TODO_

### CODE REVIEW

File: pkg/ext/stores/tokens/tokens.go [hunk 1]
```diff
@@ -1335,6 +1335,41 @@ func tokenFromSecret(secret *corev1.Secret) (*ext.Token, error) {
...
+// A value > 0 is that many milliseconds
+func ttlGreater(a, b int64) bool {
+	// Decision table
+	//
+	// a   b   | note                           | result
+	// --------+--------------------------------+-------
+	// <0  <0  | both infinite, same            | false
+	// <0  >=0 | a infinite, b not, greater     | true
+	// >=0 <0  | b infinite, a not, less        | false
+	// >=0 >=0 | map 0 to 30d, rhegular compare | t/f
```
Comment (pjbgf) [inline pkg/ext/stores/tokens/tokens.go:1351]:
```suggestion
	// >=0 >=0 | map 0 to 30d, regular compare | t/f
```

File: pkg/ext/stores/tokens/tokens.go [hunk 2]
```diff
@@ -1335,6 +1335,41 @@ func tokenFromSecret(secret *corev1.Secret) (*ext.Token, error) {
 	return token, nil
 }
 
+// ttlGreater compares the two TTL as and b. It returns true if a is greater than b.
```
Comment (pjbgf) [inline pkg/ext/stores/tokens/tokens.go:1338]:
```suggestion
// ttlGreater compares the two TTL, a and b. It returns true if a is greater than b.
```

File: pkg/ext/stores/tokens/tokens.go [hunk 3]
```diff
@@ -1335,6 +1335,41 @@ func tokenFromSecret(secret *corev1.Secret) (*ext.Token, error) {
...
+	}
+	if a < 0 && b >= 0 {
+		return true
+	}
+	if a >= 0 && b < 0 {
+		return false
+	}
+
+	if a == 0 {
+		a = ThirtyDays
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply pkg/ext/stores/tokens/tokens.go:1364]: Was not aware of the setting. Is the default indeed the max ? ...

Should be possible to integrate this into the store

```
% grep -rn auth-token-max-ttl-minutes pkg/
pkg/settings/setting.go:147:	AuthTokenMaxTTLMinutes = NewSetting("auth-token-max-ttl-minutes", "129600") // 90 days

% grep -rn AuthTokenMaxTTLMinutes pkg/
pkg/auth/tokens/manager.go:843:// ClampToMaxTTL will return the duration of the provided TTL or the duration of settings.AuthTokenMaxTTLMinutes whichever is smaller.
pkg/auth/tokens/manager.go:845:	maxTTL, err := ParseTokenTTL(settings.AuthTokenMaxTTLMinutes.Get())
pkg/auth/tokens/manager.go:847:		return 0, fmt.Errorf("failed to parse setting '%s': %w", settings.AuthTokenMaxTTLMinutes.Name, err)
pkg/settings/setting.go:146:	// AuthTokenMaxTTLMinutes is the max allowable time to live for tokens. Excluding those created for UI sessions which is controlled by AuthUserSessionTTLMinutes.
pkg/settings/setting.go:147:	AuthTokenMaxTTLMinutes = NewSetting("auth-token-max-ttl-minutes", "129600") // 90 days
```

[the remark this answers is not in the document]
Comment (pmatseykanets) [reply pkg/ext/stores/tokens/tokens.go:1364]: > Was not aware of the setting. Is the default indeed the max ?

As far as I know, yes.

[the remark this answers is not in the document]
Comment (andreas-kupries) [reply pkg/ext/stores/tokens/tokens.go:1364]: If this is setting is not only used to provide the default, but also to actually limit the max TTL (currently not done), the fact that we use a `value < 0` to represent `+infinity` becomes irrelevant. Because we then cannot specify a token which never expires. Is that ok ?

File: pkg/ext/stores/tokens/tokens.go [hunk 4]
```diff
@@ -1384,3 +1399,25 @@ func ensureNameOrGenerateName(token *ext.Token) error {
 		"Token \"%s\" is invalid: metadata.name: Required value: name or generateName is required",
 		token.ObjectMeta.Name))
 }
+
+func clampMaxTTL(millis int64) (int64, error) {
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply pkg/ext/stores/tokens/tokens.go:1403]: Ok. That is then a gateway to non-expiring tokens. I.e. with the max not set (0 or less) allow ttl < 0, +infinity. And the special comparison comes back. Will fix.

[the remark this answers is not in the document]
Comment (andreas-kupries) [reply pkg/ext/stores/tokens/tokens.go]: Done, b3d03ccf66d71a062986ebd712cd4900f0059023

File: pkg/ext/stores/tokens/tokens.go [hunk 5]
```diff
@@ -1384,3 +1395,71 @@ func ensureNameOrGenerateName(token *ext.Token) error {
...
+	maxTTL, err := tokens.ParseTokenTTL(settings.AuthTokenMaxTTLMinutes.Get())
+
+	if err != nil {
+		return 0, fmt.Errorf("failed to parse setting '%s': %w", settings.AuthTokenMaxTTLMinutes.Name, err)
+	}
+
+	return maxTTL.Milliseconds(), nil
+}
+
+// ttlGreater compares the two TTL as and b. It returns true if a is greater than b.
```
Comment (JonCrowther) [inline pkg/ext/stores/tokens/tokens.go:1439]:
```suggestion
// ttlGreater compares the two TTL a and b. It returns true if a is greater than b.
```

**Review by JonCrowther — DISMISSED** [review]
Tiny typo, otherwise lgtm

File: pkg/ext/stores/tokens/tokens.go [hunk 6]
```diff
@@ -1301,10 +1309,13 @@ func tokenFromSecret(secret *corev1.Secret) (*ext.Token, error) {
 	if err != nil {
 		return token, err
 	}
-	// inject default on retrieval
-	if ttl == 0 {
-		ttl = ThirtyDays
+
+	// clamp and inject default on retrieval
```
Comment (pmatseykanets) [inline pkg/ext/stores/tokens/tokens.go:1313]: I don't think we can "inject" the default on retrieval. The object should be immutable on read. The setting can change at any point in time resulting in a different outcome.
The default makes sense only on write - when creating and maybe when updating. Which means that the ternary (0<, 0, >0) representation of ttl in the backing secret doesn't make a lot of sense.

File: pkg/ext/stores/tokens/tokens.go [hunk 7]
```diff
@@ -1301,10 +1309,13 @@ func tokenFromSecret(secret *corev1.Secret) (*ext.Token, error) {
...
+	ttl, err = clampMaxTTL(ttl)
+	if err != nil {
+		return token, err
```
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply pkg/ext/stores/tokens/tokens.go:1316]: Checking ...

File: pkg/ext/stores/tokens/tokens.go (hunk 6 above)
Comment (andreas-kupries) [reply pkg/ext/stores/tokens/tokens.go:1313]: The stuff on retrieval has the thought of `protect against shenigans due to somebody managing to edit the backing store directly` behind it, if I remember my own thoughts correctly. 

If we drop that idea then yes, it is much better, and likely easier, to perform default resolution and clamping only on create and update, and simply take the backing store as truth to not change on list, get, watch.

File: pkg/ext/stores/tokens/tokens.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (andreas-kupries) [reply pkg/ext/stores/tokens/tokens.go:1316]: IIRC at some point in the history of the code it was useful/necessary to have access to a partially decoded token. In the code as-is nothing of that seems to remain. Will switch to idiomatic returns.

[the remark this answers is not in the document]
Comment (andreas-kupries) [reply pkg/ext/stores/tokens/tokens.go]: done 559c6951b392ab0c84f6587c3dba01164a2a4692

File: pkg/ext/stores/tokens/tokens.go (hunk 6 above)
Comment (andreas-kupries) [reply pkg/ext/stores/tokens/tokens.go]: done, a27d428eb5c09934b028d91d51882d612ad53f47
also, not sure what you looked at to still have `ThirtyDays` in it. That should have been gone in the PR.

**Review by pmatseykanets — APPROVED** [review]

**Review by JonCrowther — APPROVED** [review]

File: pkg/ext/stores/tokens/tokens.go (hunk 6 above)
Comment (andreas-kupries) [reply pkg/ext/stores/tokens/tokens.go]: Yes. I am not sure why I was thinking that way.
